### PR TITLE
✨ add card.weight

### PR DESCRIPTION
**Pull request by staff0rd**
* Adds card weighting
* Adds `validateGameDefinition`

- [x] Needs #24 
- [x] Needs #27

### CODE REVIEW

**Comment by AymericG** [comment]
Should we implement unit tests at this stage? If not when will we start?

**Comment by staff0rd** [comment]
I will add them now

File: elements/reigns/src/features/game/gameSlice.ts [hunk 1]
```diff
@@ -27,8 +29,11 @@ function setValue(statUpdate: number, stat: IStat, state: IGameState) {
   }
 }
 
+const cardsByWeight = (cards: ICard[]) =>
+  cards.flatMap((card) => range(0, card.weight).map(() => card));
```
Comment (AymericG) [inline elements/reigns/src/features/game/gameSlice.ts:33]: I feel we could avoid lodash by using [...Array(card.weight).keys()]

Comment (staff0rd) [reply elements/reigns/src/features/game/gameSlice.ts:33]: is package size a concern?

File: elements/reigns/src/features/game/gameSlice.ts [hunk 2]
```diff
@@ -88,7 +93,8 @@ export const gameSlice = createSlice({
       })
       .addCase(initializeGame.fulfilled, (state, action) => {
         state.phase = GamePhase.NOT_STARTED;
-        state.definition = action.payload;
+
+        state.definition = validateGameDefinition(action.payload);
```
Comment (AymericG) [inline elements/reigns/src/features/game/gameSlice.ts:97]: if `validateGameDefinition` throws an error I think we should make the UI show the ERROR phase.

File: elements/reigns/src/features/game/validateGameDefinition.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+): IGameDefinition => ({
+  ...definition,
+  ...validateCards(definition.cards),
+});
+const validateCards = (cards: ICard[] | undefined) => {
+  if (!cards || cards.length === 0) {
+    throw new Error("No cards found");
+  }
+
+  for (let i = 0; i < cards.length; i++) {
```
Comment (AymericG) [inline elements/reigns/src/features/game/validateGameDefinition.ts:14]: `for (const card of cards)` is less verbose I think

Comment (AymericG) [reply elements/reigns/src/features/game/validateGameDefinition.ts:14]: Oh you want the index.

File: elements/reigns/src/features/game/gameSlice.ts (hunk 2 above)
Comment (staff0rd) [reply elements/reigns/src/features/game/gameSlice.ts]: done

File: elements/reigns/src/features/game/gameSlice.ts (hunk 1 above)
Comment (staff0rd) [reply elements/reigns/src/features/game/gameSlice.ts]: b39881dd23cdf9e11ecf46f1717d8b35c10c989d

File: elements/reigns/src/features/game/gameSlice.ts [hunk 4]
```diff
@@ -27,8 +28,12 @@ function setValue(statUpdate: number, stat: Stat, state: GameState) {
   }
 }
 
+const cardsByWeight = (cards: Card[]) =>
```
Comment (maxired) [inline elements/reigns/src/features/game/gameSlice.ts:31]: given this function name, I would expect an object whose keys being weight

Comment (staff0rd) [reply elements/reigns/src/features/game/gameSlice.ts]: b65cbac032d29e5922adb975199216a91c531e66

**Review by maxired — APPROVED** [review]

**Pull request merged**